test/libsolidity/syntaxTests/constructor/constructor_function_parameter_err.sol

```diff
@@ -0,0 +1,36 @@
+contract D {
+    constructor(function() external pure returns (uint) ) {
+    }
+}
+
+contract C {
+    function f() public returns (uint r) {
+        // An assert used to fail the function types are not exactly equal (pure, view) v/s default
+        // ok
+        D d1 = new D(this.testPure);
+        // not okay
+        D d2 = new D(this.testView);
+        // not okay
+        D d3 = new D(this.testDefault);
+        // not okay
+        D d4 = new D(this.testDifferentSignature);
```

Comment:
I'd rename these two tests to something like `constructor_function_parameter_allowed_conversions.sol` and `constructor_function_parameter_disallowed_conversion.sol` for clarity. When I'm checking if there is already a test for something the name is usually all I see so I think good naming for this is important.

Also, I took a look at the syntax tests we have and it seems that passing external functions in parameters has almost no coverage (unless some unrelated tests do that but that's very hard to check). So I think it would be good to add some more cases while you're at it. Here's some ideas:
- More cases that should now be allowed:
    - Passing in an external library function with different mutability that can be implicitly converted.
- More cases that should still be disallowed:
    - Passing in an internal function.
    - Passing in a bound function.
    - Passing in a built-in function.
- Does your change allow the implicit conversion in other places where stuff gets ABI-encoded?
    - In parameters of external contract and library functions?
    - In return values of external functions and public state variable getters?
    - In `abi.decode()`/`abi.encode()`?

Reply:
> Passing in an external library function with different mutability that can be implicitly converted.

What would be test case for this?

Reply:
Something like this:

```solidity
library L {
    function f(uint, bytes memory) public pure {}
}

contract C {
    constructor(function (uint, bytes memory) external) {}
}
```

Basically I just meant checking this with library functions because they have some extra rules (e.g. you can't have a pointer to one that takes storage arguments).

Reply:
This would be a test for the disallowed category, right?

Reply:
Yeah. Well, initially I thought that you could have such a pointer as long as the library function has a normal signature (e.g. does not do things disallowed in contracts like accepting storage parameters) but I remembered wrong. Pointers to external library functions are not allowed.

And I see that your updated test already covers this case.